algo/src/main/java/org/neo4j/graphalgo/linkprediction/LinkPrediction.java

```diff
@@ -0,0 +1,136 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * <p>
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.graphalgo.linkprediction;
+
+import org.neo4j.graphalgo.core.ProcedureConfiguration;
+import org.neo4j.graphdb.*;
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.procedure.Context;
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Name;
+import org.neo4j.procedure.UserFunction;
+
+import java.util.*;
+
+public class LinkPrediction {
+
+    @Context
+    public GraphDatabaseAPI api;
+
+    @UserFunction("algo.linkprediction.adamicAdar")
+    @Description("algo.linkprediction.adamicAdar(node1:Node, node2:Node, {relationshipQuery:'relationshipName', direction:'BOTH'}) " +
+            "given two nodes, calculate Adamic Adar similarity")
+    public double adamicAdarSimilarity(@Name("node1") Node node1, @Name("node2") Node node2,
+                                       @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+        // https://en.wikipedia.org/wiki/Adamic/Adar_index
+
+        if (node1 == null || node2 == null) {
+            throw new RuntimeException("Nodes must not be null");
+        }
+
+        ProcedureConfiguration configuration = ProcedureConfiguration.create(config);
+        RelationshipType relationshipType = configuration.getRelationship();
+        Direction direction = configuration.getDirection(Direction.BOTH);
+
+        List<Number> vector = new ArrayList<>();
+        try(Transaction tx = api.beginTx()) {
+            Set<Node> neighbors = findPotentialNeighbors(node1, relationshipType, direction);
+            neighbors.removeIf(node -> noCommonNeighbors(node, relationshipType, direction, node2));
+            neighbors.forEach(neighbor -> vector.add(degree(relationshipType, direction, neighbor)));
+            tx.success();
+        }
+
+        return sumInverseLog(vector);
+    }
+
+    @UserFunction("algo.linkprediction.resourceAllocation")
+    @Description("algo.linkprediction.resourceAllocation(node1:Node, node2:Node, {relationshipQuery:'relationshipName', direction:'BOTH'}) " +
+            "given two nodes, calculate Resource Allocation similarity")
+    public double resourceAllocationSimilarity(@Name("node1") Node node1, @Name("node2") Node node2,
+                                       @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+        ProcedureConfiguration configuration = ProcedureConfiguration.create(config);
+        RelationshipType relationshipType = configuration.getRelationship();
+        Direction direction = configuration.getDirection(Direction.BOTH);
+
+        List<Number> vector = new ArrayList<>();
+        try(Transaction tx = api.beginTx()) {
```

Comment:
why a new tx?

Reply:
I always forget that functions have an implicit tx!